drivers/sdhc/sdhci_common.c

```diff
@@ -0,0 +1,794 @@
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int sdhci_set_uhs_timing(struct sdhci_common *sdhci_ctx, enum sdhc_timing_mode timing)
+{
+	bool hspd = false;
+	uint16_t reg;
+
+	reg = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+	reg &= ~SDHCI_CTRL_UHS_MASK;
+
+	switch (timing) {
+	case SDHC_TIMING_LEGACY:
+	case SDHC_TIMING_SDR12:
+		reg |= SDHCI_CTRL_UHS_SDR12;
+		break;
+	case SDHC_TIMING_SDR25:
+		reg |= SDHCI_CTRL_UHS_SDR25;
+		hspd = true;
+		break;
+	case SDHC_TIMING_SDR50:
+		reg |= SDHCI_CTRL_UHS_SDR50;
+		break;
+	case SDHC_TIMING_DDR50:
+		reg |= SDHCI_CTRL_UHS_DDR50;
+		break;
+	case SDHC_TIMING_SDR104:
+		reg |= SDHCI_CTRL_UHS_SDR104;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	sdhci_enable_sd_clock(sdhci_ctx, false);
+
+	LOG_DBG("sdhc:timing: SDHCI_HOST_CONTROL2: %04x", reg);
+	sys_write16(reg, sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+	sdhci_set_hspd(sdhci_ctx, hspd);
+
+	sdhci_enable_sd_clock(sdhci_ctx, true);
+	return 0;
+}
+
+static int sdhc_stop_transfer(struct sdhci_common *sdhci_ctx)
+{
+	struct sdhc_command hdc_cmd = {0};
+
+	hdc_cmd.opcode = SD_STOP_TRANSMISSION;
+	hdc_cmd.arg = 0;
+	hdc_cmd.response_type = SD_RSP_TYPE_R1;
+	hdc_cmd.timeout_ms = 1000;
+
+	return sdhci_send_cmd(sdhci_ctx, &hdc_cmd, false);
+}
+
+static bool sdhc_need_cmd12(struct sdhc_command *cmd)
+{
+	return (cmd->opcode == SD_READ_MULTIPLE_BLOCK || cmd->opcode == SD_WRITE_MULTIPLE_BLOCK);
+}
+
+static int sdhci_poll_data_complete(struct sdhci_common *sdhci_ctx, int32_t timeout)
+{
+	uint32_t reg_normal_int_stat = 0;
+	int ret = -ETIMEDOUT;
+
+	timeout = timeout ? timeout : 1;
+	timeout = timeout * 100;
+
+	while (timeout > 0) {
+		reg_normal_int_stat = sys_read32(sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		if (reg_normal_int_stat & SDHCI_INT_XFER_COMPLETE) {
+			sys_write32(SDHCI_INT_XFER_COMPLETE,
+				    sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+			ret = 0;
+			break;
+		}
+
+		k_busy_wait(10);
+		timeout--;
+	}
+
+	if (reg_normal_int_stat & SDHCI_INT_ERROR) {
+		LOG_ERR("sdhc:req: data xfer error int_status:%08x", reg_normal_int_stat);
+		sys_write32(SDHCI_INT_ERROR_MASK, sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		ret = -EIO;
+	}
+
+	if (!timeout) {
+		LOG_ERR("sdhc:req: data xfer timeout int_status:%08x", reg_normal_int_stat);
+	}
+
+	return ret;
+}
+
+static void sdhci_data_port_io(struct sdhci_common *sdhci_ctx, uint8_t *data, uint32_t size,
+			       bool read)
+{
+	uint32_t *offs = (uint32_t *)data;
+	int i;
+
+	for (i = 0; i < size; i += 4, offs++) {
+		if (read) {
+			*offs = sys_read32(sdhci_ctx->reg_base + SDHCI_BUFFER);
+		} else {
+			sys_write32(*offs, sdhci_ctx->reg_base + SDHCI_BUFFER);
+		}
+	}
+}
+
+static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data *data,
+			       bool is_read)
+{
+	uint32_t block = data->blocks;
+	uint32_t reg_present_state;
+	uint32_t int_status, rdy_mask;
+	uint8_t *data_ptr = data->data;
+	int32_t timeout;
+
+	timeout = data->timeout_ms ? data->timeout_ms : 1;
```

Comment:
looks like timeout_ms for data and cmd can be negative and be equal to `SDHC_TIMEOUT_FOREVER`

Reply:
FOREVER not supported for polling - plan was to support for IRQ.

Reply:
Why not? Is there a timeout error in this case returned by host controller, and do you leave the loop?

Reply:
Created "TODO: consider proper handling of FOREVER timeout in poling mode"